src/kafka/producer/metadata_manager.cc

```diff
@@ -38,9 +40,34 @@ namespace kafka {
         });
     }
 
-    metadata_response& metadata_manager::get_metadata() {
+    seastar::future<> metadata_manager::refresh_coroutine(std::chrono::seconds dur) {
+        return seastar::async({}, [this, dur]{
+            while(_keep_refreshing) {
+                refresh_metadata().get();
+                try {
+                    seastar::sleep_abortable(dur, _stop_refresh).get();
+                } catch (seastar::sleep_aborted e) {}
+            }
+            _refresh_finished.signal();
+            return;
+        });
+    }
+
+    metadata_response &metadata_manager::get_metadata() {
```

Comment:
Is a reference here entirely safe? I wonder if there can be a situation where part of the info is taken from the old metadata and then the other part from the new one that gets set (after some refresh happened in the background)

Reply:
It's not, fixed it